### PR TITLE
Gradients for inverse trig functions (arccos, arcsin, arctan)

### PR DESCRIPTION
**Pull request by shak360**
This is for progress in issue #161  .

- Added gradients for `arccos` and `arcsin`.
- Added a test that checks the gradients of those functions.
- Moved some grad functions around to be closer to each other.

### CODE REVIEW

File: nx/test/nx/defn/grad_test.exs [hunk 1]
```diff
@@ -478,6 +478,19 @@ defmodule Nx.Defn.GradTest do
     end
   end
 
+  describe "inverse trig family" do
+    defn grad_arcsin(t), do: grad(t, Nx.arcsin(t))
+    defn grad_arccos(t), do: grad(t, Nx.arccos(t))
+
+    test "computes gradient of inverse trig functions" do
+      for _ <- 1..100 do
+        t = Nx.random_uniform({}, -0.999, 0.999, type: {:f, 64})
```
Comment (shak360) [inline nx/test/nx/defn/grad_test.exs:487]: The domain of these functions is between -1 and 1, noninclusive.

**Comment by seanmor5** [comment]
@shak360 Hey thanks for the contribution and awesome work! If you want to you can also add `arctan`: `1/ 1 + x^2`

I understand it's domain is -inf - +inf so you might want to test that differently, but it's trivial enough that I'm not as concerned with needing to test over that. Let me know!

**Comment by shak360** [comment]
> @shak360 Hey thanks for the contribution and awesome work! If you want to you can also add `arctan`: `1/ 1 + x^2`
> 
> I understand it's domain is -inf - +inf so you might want to test that differently, but it's trivial enough that I'm not as concerned with needing to test over that. Let me know!

Sounds good! I've added the arctan gradient and the test.

File: nx/lib/nx/defn/grad.ex [hunk 2]
```diff
@@ -406,13 +406,18 @@ defmodule Nx.Defn.Grad do
...
 
-  defp grad(:tanh, [x], ans, g, cache) do
-    g = Nx.multiply(g, Nx.subtract(1.0, Nx.multiply(ans, ans)))
+  defp grad(:cos, [x], _ans, g, cache) do
+    g = Nx.multiply(g, Nx.negate(Nx.sin(x)))
+    to_grad(x, g, cache)
+  end
+
+  defp grad(:arccos, [x], _ans, g, cache) do
+    g = Nx.divide(g, Nx.multiply(-1.0, Nx.power(Nx.subtract(1.0, Nx.power(x, 2.0)), 0.5)))
```
Comment (seanmor5) [inline nx/lib/nx/defn/grad.ex:420]: This is me just nitpicking:

```suggestion
    g = Nx.multiply(g, Nx.negate(Nx.rsqrt(Nx.subtract(1.0, Nx.power(x, 2.0)))))
```

The `negate` is a bit more deliberate so when you're inspecting a complex expression it's harder to miss what's going on. I think a compiler has a better chance of optimizing `rsqrt` versus the `divide(...)` pattern, but that could be wrong. Either way it's also a bit more deliberate, so harder to miss in debugging :)

File: nx/lib/nx/defn/grad.ex [hunk 3]
```diff
@@ -406,13 +406,18 @@ defmodule Nx.Defn.Grad do
     to_grad(x, g, cache)
   end
 
-  defp grad(:sqrt, [x], ans, g, cache) do
-    g = Nx.multiply(g, Nx.divide(0.5, ans))
+  defp grad(:arcsin, [x], _ans, g, cache) do
+    g = Nx.divide(g, Nx.power(Nx.subtract(1.0, Nx.power(x, 2.0)), 0.5))
```
Comment (seanmor5) [inline nx/lib/nx/defn/grad.ex:410]:
```suggestion
    g = Nx.divide(g, Nx.rsqrt(Nx.subtract(1.0, Nx.power(x, 2.0))))
```

**Review by seanmor5 — COMMENTED** [review]
Just two nitpicks, they really don't make much of a difference :) Otherwise great job and thanks for the contribution!

File: nx/test/nx/defn/grad_test.exs [hunk 4]
```diff
@@ -478,6 +478,22 @@ defmodule Nx.Defn.GradTest do
...
   end
 
+  describe "inverse trig family" do
+    defn grad_arcsin(t), do: grad(t, Nx.arcsin(t))
+    defn grad_arccos(t), do: grad(t, Nx.arccos(t))
+    defn grad_arctan(t), do: grad(t, Nx.arctan(t))
+
+    test "computes gradient of inverse trig functions" do
+      for _ <- 1..100 do
+        t = Nx.random_uniform({}, -0.996, 0.996, type: {:f, 64})
```
Comment (shak360) [inline nx/test/nx/defn/grad_test.exs:488]: @seanmor5 I added the changes in the grad functions for `arccos` and `arcsin` but now the tests only work up to plus/minus 0.996. I see the benefit of having the more readable gradient though.

File: nx/lib/nx/defn/grad.ex [hunk 5]
```diff
@@ -406,13 +406,18 @@ defmodule Nx.Defn.Grad do
...
 
-  defp grad(:tanh, [x], ans, g, cache) do
-    g = Nx.multiply(g, Nx.subtract(1.0, Nx.multiply(ans, ans)))
+  defp grad(:cos, [x], _ans, g, cache) do
+    g = Nx.multiply(g, Nx.negate(Nx.sin(x)))
+    to_grad(x, g, cache)
+  end
+
+  defp grad(:arccos, [x], _ans, g, cache) do
+    g = Nx.multiply(g, Nx.negate(Nx.rsqrt(Nx.subtract(1.0, Nx.power(x, 2.0)))))
```
Comment (shak360) [inline nx/lib/nx/defn/grad.ex:420]: Gradient using `Nx.negate` and `Nx.rsqrt`.

File: nx/lib/nx/defn/grad.ex [hunk 6]
```diff
@@ -406,13 +406,18 @@ defmodule Nx.Defn.Grad do
     to_grad(x, g, cache)
   end
 
-  defp grad(:sqrt, [x], ans, g, cache) do
-    g = Nx.multiply(g, Nx.divide(0.5, ans))
+  defp grad(:arcsin, [x], _ans, g, cache) do
+    g = Nx.multiply(g, Nx.rsqrt(Nx.subtract(1.0, Nx.power(x, 2.0))))
```
Comment (shak360) [inline nx/lib/nx/defn/grad.ex:410]: Gradient using `Nx.rsqrt`.

**Review by josevalim — APPROVED** [review]

**Pull request merged**